File: 15_Gram-Schmidt/gram-schmidt.py
# Computational Linear Algebra Ep#15 Gram-Schmidt Process
# By: Nick Space Cowboy

import numpy as np

def gram_schmidt(A):
    m, n  = A.shape
    Q = np.zeros((m, n))
    Q[:, 0] = A[:, 0] / np.linalg.norm(A[:, 0], 2)
    for i in range(1, n):
        Q[:, i] = A[:, i]
        for j in range(0, i):
            inner = np.dot(Q[:, j].T, Q[:, i])
            Q[:, i] = Q[:, i] - np.dot(inner, Q[:, j])
        Q[:, i] = Q[:, i] / np.linalg.norm(Q[:, i], 2)
    return Q


if __name__ == "__main__":
    A = np.array(np.random.randint (0, 10, (3, 3)), dtype=np.float64)
    print("A: ")
    print(A)
    print("Q: ")
    Q = gram_schmidt(A)
    print(Q)
    I = np.dot(Q.T, Q)
    print("Check: ")
    print(I)
    print(np.round(I))
    

File: Unit3/Gram-Schmidt/gram-schmidt.m
# Computational Linear Algebra Ep#15 Gram-Schmidt Process
# By: Nick Space Cowboy

clear all;
close all;

function Q = gram_schmidt(A)
    [n, k] = size(A);
    Q = zeros(n,k);
    Q(:,1) = A(:,1) / norm(A(:,1));
    for i = 2:k
        Q(:,i) = A(:,i);
        for j = 1:i-1
            Q(:,i) = Q(:,i) - (Q(:,j)'* Q(:,i)) * Q(:,j);
        end
        Q(:,i) = Q(:,i) / norm(Q(:,i));
    end
endfunction

function inner_products(M)
    [m, n] = size(M);
    for i = 2:n
        inner_product = dot(M(:,i-1), M(:, i))
    end
    inner_product = dot(M(:, 1), M(:, n))
endfunction

A = randi(10, 3, 3)
inner_products(A)
Q = gram_schmidt(A)
inner_products(Q)
check = Q * Q'
